cosmwasm/enclaves/shared/contract-engine/src/contract_validation.rs

```diff
@@ -675,24 +1151,167 @@ fn verify_message_params(
+    denom.starts_with(&voucher_prefix)
+}
+
+/// GetDenomPrefix returns the receiving denomination prefix
+fn get_denom_prefix(port_id: &str, channel_id: &str) -> String {
+    format!("{}/{}/", port_id, channel_id)
+}
+
+/// DenomTrace contains the base denomination for ICS20 fungible tokens and the
+/// source tracing information path.
+struct DenomTrace {
+    /// path defines the chain of port/channel identifiers used for tracing the
+    /// source of the fungible token.
+    path: String,
+    /// base denomination of the relayed fungible token.
+    base_denom: String,
+}
+
+impl DenomTrace {
+    /// Hash returns the hex bytes of the SHA256 hash of the DenomTrace fields using the following formula:
+    /// hash = sha256(tracePath + "/" + baseDenom)
+    pub fn hash(&self) -> Vec<u8> {
+        let hash = Sha256::digest(self.get_full_denom_path().as_bytes());
+        hash.to_vec()
+    }
+
+    /// IBCDenom a coin denomination for an ICS20 fungible token in the format
+    /// 'ibc/{hash(tracePath + baseDenom)}'. If the trace is empty, it will return the base denomination.
+    pub fn ibc_denom(&self) -> String {
+        if !self.path.is_empty() {
+            format!("ibc/{}", hex::encode(self.hash()))
+        } else {
+            self.base_denom.clone()
+        }
+    }
+
+    /// GetFullDenomPath returns the full denomination according to the ICS20 specification:
+    /// tracePath + "/" + baseDenom
+    /// If there exists no trace then the base denomination is returned.
+    pub fn get_full_denom_path(&self) -> String {
+        if self.path.is_empty() {
+            self.base_denom.clone()
+        } else {
+            self.get_prefix() + &self.base_denom
+        }
+    }
+
+    // GetPrefix returns the receiving denomination prefix composed by the trace info and a separator.
+    fn get_prefix(&self) -> String {
+        return format!("{}/", self.path);
+    }
+}
+
+/// ParseDenomTrace parses a string with the ibc prefix (denom trace) and the base denomination
+/// into a DenomTrace type.
+///
+/// Examples:
+///
+/// - "portidone/channel-0/uatom" => DenomTrace{Path: "portidone/channel-0", BaseDenom: "uatom"}
+/// - "portidone/channel-0/portidtwo/channel-1/uatom" => DenomTrace{Path: "portidone/channel-0/portidtwo/channel-1", BaseDenom: "uatom"}
+/// - "portidone/channel-0/gamm/pool/1" => DenomTrace{Path: "portidone/channel-0", BaseDenom: "gamm/pool/1"}
+/// - "gamm/pool/1" => DenomTrace{Path: "", BaseDenom: "gamm/pool/1"}
+/// - "uatom" => DenomTrace{Path: "", BaseDenom: "uatom"}
+fn parse_denom_trace(raw_denom: &str) -> DenomTrace {
+    let denom_split: Vec<&str> = raw_denom.split('/').collect();
+
+    if denom_split.len() == 1 {
+        return DenomTrace {
+            path: "".to_string(),
+            base_denom: raw_denom.to_string(),
+        };
+    }
+
+    let (path, base_denom) = extract_path_and_base_from_full_denom(&denom_split);
+
+    DenomTrace { path, base_denom }
```

Comment:
Same as above

Reply:
This also applies to the rest of the random code here

Reply:
https://github.com/scrtlabs/SecretNetwork/commit/d487b138c